### PR TITLE
[enhancement] add ocb custom collector support for ddflareextension

### PR DESCRIPTION
**Pull request by jackgopack4**
### What does this PR do?
enables ddflareextension to be built with OpenTelemetry Collector Builder (ocb) and included with an open-source collector

### Motivation
closes https://datadoghq.atlassian.net/browse/OTEL-2148

### Describe how to test/QA your changes
- make directory `/tmp/otel-ci/` (e.g. `mkdir /tmp/otel-ci/`)
- download ocb for your platform, version v0.105.0 is confirmed to work: https://github.com/open-telemetry/opentelemetry-collector-releases/releases/tag/cmd%2Fbuilder%2Fv0.111.0 (e.g. `wget -O /tmp/otel-ci/ocb https://github.com/open-telemetry/opentelemetry-collector-releases/releases/download/cmd%2Fbuilder%2Fv0.111.0/ocb_0.111.0_darwin_arm64`)
- make this binary executable (e.g. `chmod +x /tmp/otel-ci/ocb`)
- run the following command: `/tmp/otel-ci/ocb --config=/path/to/datadog-agent/test/otel/testdata/builder-config.yaml`
- ensure the above command completes successfully (no error messages shown)
- run the built collector binary located at /tmp/otel-ci/otelcol-custom/otelcol-custom (e.g. `/tmp/otel-ci/otelcol-custom/otelcol-custom --config /path/to/datadog-agent/test/otel/testdata/collector-config.yaml`)
- open a new terminal window/tab and send the following request: `curl -k https://localhost:7777`)
- ensure there are no error messages shown
- go back to the other terminal window and terminate the collector process (ctrl-C)

### Possible Drawbacks / Trade-offs
Due to some of the components of ddflare testing framework requiring an older version of `datadogconnector`, this pulls in an older version of another datadog agent module that is no longer updated (`comp/core/log`). Once we are able to split OTel-Agent into its own independent Go module (not dependent on the go.mod/imports of the parent datadog-agent module), we will be able to update these dependencies and remove these references. For now I needed to update `modules.py` tests to explicitly exclude this module and the dependencies that it imports, since they are no longer maintained  as `go.mod` files within Agent codebase. I welcome the feedback and input of the Agent DevX folks if there is a better way to do this, or if I need to file a ticket to make sure that this work does get done after we separate OTel-Agent in the future.

### Additional Notes

It could be simpler to split off this into a separate implementation such as `ddflareextension/impl-ocb` but there is a lot more shared code than not; if someone more familiar with the agent component model has strong opinions about this I would be willing to consider that. There's a lot of shared code and only a few methods that needed to be changed/added to add ocb support, so I don't know if it would be worth duplicating code to achieve that.

### CODE REVIEW

File: comp/otelcol/converter/impl/go.mod
Comment (jackgopack4) [inline comp/otelcol/converter/impl/go.mod:1]: module is not used here and causes issues with linting/go mod tidy checks

File: go.mod
Comment (jackgopack4) [inline go.mod:1]: building ddflareextension/impl as a separate package requires minimum versions of go modules that are published on go.dev; many of these version numbers were not actually published versions

**Comment by jade-guiton-dd** [comment]
LGTM, but I'll leave it to Dinesh for the opentelemetry team approval, as I don't have the context for some of the decisions made here.

**Review by KevinFairise2 — APPROVED** [review]

**Comment by jackgopack4** [comment]
will likely have some merge conflicts after [`a90ab38` (#30061)](https://github.com/DataDog/datadog-agent/pull/30061/commits/a90ab38e60479d78c14e6ae43cc70a419f47aeed) is merged in; will wait to rebase/update after that PR is merged

File: .gitlab/integration_test/otel.yml
Comment (CelianR) [inline .gitlab/integration_test/otel.yml:70]:
```suggestion
```

Is this useful?

File: .gitlab/integration_test/otel.yml
Comment (CelianR) [inline .gitlab/integration_test/otel.yml:86]:
```suggestion
    - echo 'Building collector with OCB and test ddflare extension'
    - /tmp/otel-ci/ocb --config=/tmp/otel-ci/builder-config.yaml > ocb-output.log 2>&1
    - grep -q 'Compiled' ocb-output.log || (echo "OCB failed to compile" && exit 1)
    - grep -q '{"binary": "/tmp/otel-ci/otelcol-custom/otelcol-custom"}' ocb-output.log || (echo "OCB failed to compile" && exit 1)
    - /tmp/otel-ci/otelcol-custom/otelcol-custom --config /tmp/otel-ci/collector-config.yaml > otelcol-custom.log 2>&1 &
    - OTELCOL_PID=$!  # Capture the process ID
    - sleep 10  # Wait for the process to start
    - grep -q 'Everything is ready. Begin running and processing data.' otelcol-custom.log || (echo "custom collector failed to start" && kill $OTELCOL_PID && exit 1)
    - curl -k https://localhost:7777 > flare-info.log 2>&1 # get the dd flare info
    - grep -q '"provided_configuration": "ConfigProvider not supported, only enhanced config available"' flare-info.log || (echo "provided config should not be supported with ocb" && kill $OTELCOL_PID && exit 1)
    - grep -q 'extensions:\\n  - ddflare\\n' flare-info.log || (echo "ddflare extension should be enabled" && kill $OTELCOL_PID && exit 1)
```

Some formatting

**Review by CelianR — APPROVED** [review]

File: comp/otelcol/ddflareextension/impl/configstore.go
Comment (mackjmr) [inline comp/otelcol/ddflareextension/impl/configstore.go:23]: You should be able to remove all this once #30061 is merged and you pull in main 🙂

Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/configstore.go:23]: I plan to keep the providedConfigSupported so that we can still use logic to return a "unsupported" or nil value in the provided config field but yes thank you for that, I appreciate it

File: .gitlab/integration_test/otel.yml
Comment (jackgopack4) [reply .gitlab/integration_test/otel.yml:85]: iirc there was an issue with escaping of quotations/brackets in the yaml before I did this; I'll commit this and make sure it runs successfully

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/extension.go:61]: Focus on factories being nil rather than location of the call

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/extension.go:231]: this will be collapsed a little bit since EffectiveConfig will actually go away once mackenzie's changes from [`a90ab38` (#30061)](https://github.com/DataDog/datadog-agent/pull/30061/commits/a90ab38e60479d78c14e6ae43cc70a419f47aeed) are merged in

**Review by truthbk — COMMENTED** [review]
Looks good, small nits, but generally looking almost ready.

File: tasks/modules.py
Comment (jackgopack4) [reply tasks/modules.py:315]: the older version of datadogexporter or datadogconnector that was pinned in the codebase; this will be fixed after the k8s dependencies is merged bumping to 0.111.0

File: comp/otelcol/ddflareextension/impl/configstore.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/configstore.go]: It could just as easily be part of the extension, I don't have strong opinions about this one.

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/extension.go]: this is removed in favor of using Confmap in the configstore

File: .gitlab/integration_test/otel.yml
Comment (jackgopack4) [reply .gitlab/integration_test/otel.yml:77]: I wanted to check the output of the collector with this one, so I could do a loop of grep with shorter sleep and break if the phrase is found. If I skip this step I could just try curling 7777 until it succeeds or times out, that approach is probably fine

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/extension.go]: understood, fixed

**Comment by jackgopack4** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by jackgopack4** [comment]
@dineshg13 @truthbk resolved most of the concerns (still open to tweaking the integration test if that's necessary) and also added forward compatibility for v0.112.0 (two extension interface methods were renamed)

**Review by pducolin — APPROVED** [review]

**Review by davidor — APPROVED** [review]
No review needed from container-platform. It's just a dependency bump :+1:

File: comp/otelcol/ddflareextension/impl/server.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/server.go:120]: cc @truthbk are we ok with this ?

File: comp/otelcol/ddflareextension/impl/factory.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/factory.go:85]: should we do this?

File: comp/otelcol/ddflareextension/impl/factory.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/factory.go:50]: remove or handle it ?

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/extension.go:134]: no need to set false . That is default value.  We can remove this if

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/extension.go:199]: remove else . "" is zero value

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (dineshg13) [inline comp/otelcol/ddflareextension/impl/extension.go:190]: I don't see a need for is ProvidedConfig supporter. Lets get 	`customer, err = ext.configStore.getProvidedConfAsString()` , as long as it isn't empty string we have provided config

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: comp/otelcol/ddflareextension/impl/server.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/server.go:120]: added a card for backlog https://datadoghq.atlassian.net/browse/OTEL-2226

File: comp/otelcol/ddflareextension/impl/factory.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/factory.go:85]: I can remove the extra methods `Stability` and `Create` in this PR and just leave a TODO note that these methods will be deprecated with 112.0 and above, but this seemed safer because the error messages I received in the ddflare ocb integration test weren't super obvious as to what was going on

Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/factory.go:85]: or just update the message that we don't need to remove the old methods

**Review by dineshg13 — APPROVED** [review]
Approved with couple of comments

File: comp/otelcol/ddflareextension/impl/extension.go
Comment (jackgopack4) [reply comp/otelcol/ddflareextension/impl/extension.go:188]: That function doesn't return any errors only nil. Could refactor the called function to actually throw error on error

**Pull request merged**
